File: src/main/java/com/pocketempire/config/UnitNamesLoader.java
package com.pocketempire.config;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.InputStream;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

public class UnitNamesLoader {
    private static final String NAMES_FILE = "/name_pools.json";
    private static NamesConfig config;
    private static final Set<String> usedNames = new HashSet<>();
    private static final Random random = new Random();

    static {
        loadNames();
    }

    private static void loadNames() {
        ObjectMapper mapper = new ObjectMapper();
        try (InputStream is = UnitNamesLoader.class.getResourceAsStream(NAMES_FILE)) {
            config = mapper.readValue(is, NamesConfig.class);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static String getRandomName() {
        if (config == null) return "Unknown Unit";

        if (!config.first_names.isEmpty()) {
            String name = config.first_names.remove(random.nextInt(config.first_names.size()));
            usedNames.add(name);
            return name;
        }

        if (config.prefixes != null && config.suffixes != null
                && !config.prefixes.isEmpty() && !config.suffixes.isEmpty()) {
            for (int attempt = 0; attempt < 100; attempt++) {
                String name = config.prefixes.get(random.nextInt(config.prefixes.size()))
                        + config.suffixes.get(random.nextInt(config.suffixes.size()));
                if (usedNames.add(name)) {
                    return name;
                }
            }
        }

        return "Unknown Unit";
    }
}

File: src/main/kotlin/com/pocketempire/config/UnitNamesLoader.kt
package com.pocketempire.config

import com.fasterxml.jackson.databind.ObjectMapper
import java.util.Random

object UnitNamesLoader {

    private const val NAMES_FILE = "/name_pools.json"
    private var config: NamesConfig? = null
    private val usedNames = HashSet<String>()
    private val random = Random()

    init {
        loadNames()
    }

    private fun loadNames() {
        try {
            val inputStream = requireNotNull(
                UnitNamesLoader::class.java.getResourceAsStream(NAMES_FILE)
            ) {
                "Could not find $NAMES_FILE"
            }

            inputStream.use {
                config = ObjectMapper().readValue(it, NamesConfig::class.java)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun getRandomName(): String {
        val namesConfig = config ?: return "Unknown Unit"

        if (namesConfig.first_names.isNotEmpty()) {
            val name = namesConfig.first_names.removeAt(random.nextInt(namesConfig.first_names.size))
            usedNames.add(name)
            return name
        }

        if (namesConfig.prefixes.isNotEmpty() && namesConfig.suffixes.isNotEmpty()) {
            repeat(100) {
                val name = namesConfig.prefixes[random.nextInt(namesConfig.prefixes.size)] +
                    namesConfig.suffixes[random.nextInt(namesConfig.suffixes.size)]
                if (usedNames.add(name)) {
                    return name
                }
            }
        }

        return "Unknown Unit"
    }

    @JvmStatic
    fun getRandomCityName(): String {
        val namesConfig = config ?: return "New City"

        if (namesConfig.city_names.isEmpty()) {
            return "New City"
        }

        val name = namesConfig.city_names.removeAt(random.nextInt(namesConfig.city_names.size))
        usedNames.add(name)
        return name
    }
}
